src/item_collaborative_filtering.py

```diff
@@ -0,0 +1,146 @@
+"""
+Item to Item Collaborative Filtering
+Paper: https://www.computer.org/csdl/mags/ic/2017/03/mic2017030012.pdf
+"""
+from itertools import permutations
+
+from joblib import Parallel, delayed
+import numpy as np
+import pandas as pd
+
+
+class ItemItemCollaborativeFiltering:
+
+    def __init__(self, item_column='item_id', user_column='user_id'):
+        self.item_column = item_column
+        self.user_column = user_column
+
+    def __generate_item_pairs(self, df, item):
+        """
+        Creates permutations/pairs of two products
+        :param df: dataframe with columns [user_id, item_id]
+        :param item: item that is paired with all other items, if None - all possible pairs are created
+        :return: list of tuples, length=df['item'].nunique()
+        """
+        if item is not None:
+            return [(item, paired_item) for paired_item in df[self.item_column].unique() if paired_item!=item]
+        else:
+            return [(item, paired_item) for
+                    item, paired_item in permutations(df[self.item_column].unique(), 2) if paired_item!=item]
+
+
+    def __count_common_item_pair_users(self, df, item_pair):
+        """
+        Computes the number of users that interacted (e.g. watched, purchased) BOTH with item1 and item2
+        :param df: dataframe with columns [user_id, item_id]
+        :param item_pair: tuple(item1, item2)
+        :return: int, number of users in common for the item pair (item1, item2)
+        """
+        item1, item2 = item_pair[0], item_pair[1]
+        item1_users = set(df.loc[df[self.item_column]==item1, self.user_column].unique())
+        item2_users = set(df.loc[df[self.item_column]==item2, self.user_column].unique())
+
+        common_users = len(item1_users.intersection(item2_users))
+
+        return item_pair, common_users
```

Comment:
Changed the variable to `common_users_count`.
I returned `item_pair` because in the method `fit_recommendations()` line 134 I'm filtering item pairs with `common_users_count  > 0`.

Reply:
> I returned item_pair because in the method fit_recommendations() line 134 I'm filtering item pairs with common_users_count > 0.

Ah sorry, missed that :)